src/psyclone/domain/lfric/psyir.py

```diff
@@ -0,0 +1,162 @@
+     "precision": "i_def", "vars": [], "short_name": "nqp_v"},
+    {"name": "number of qr points", "intrinsic": "integer",
+     "precision": "i_def", "vars": [], "short_name": "nqp"}]
+
+for info in scalar_datatypes:
+    NAME = "".join(info["name"].title().split())
+    INTRINSIC = info["intrinsic"].upper()
+    PRECISION = info["precision"]
+    ARGS = ["self", "name"] + info["vars"] + ["interface=None"]
+    VARS = ["        self.{0} = {0}".format(var) for var in info["vars"]]
+    exec('''
+class {0}DataType(ScalarType):
+    def __init__(self):
+        super({0}DataType, self).__init__(
+            ScalarType.Intrinsic.{1}, {2})
+    '''.format(NAME, INTRINSIC, PRECISION.upper()))
+    exec('''
+class {0}DataSymbol(DataSymbol):
+    def __init__({1}):
+{2}
+        super({0}DataSymbol, self).__init__(
+            name, {0}DataType(), interface=interface)
+    '''.format(NAME, ", ".join(ARGS), "\n".join(VARS)))
+
+field_datatypes = [
+    {"name": "real field data", "scalar_type": "lfric real scalar",
+     "dims": ["number of unique dofs"], "vars": ["fs"]},
+    {"name": "integer field data", "scalar_type": "lfric integer scalar",
+     "dims": ["number of unique dofs"], "vars": ["fs"]},
+    {"name": "logical field data", "scalar_type": "lfric logical scalar",
+     "dims": ["number of unique dofs"], "vars": ["fs"]}]
+
+# TBD: #xxx Sort out how to deal with multi-dimensional array
+# information. (diff) basis function dimensions change depending on
+# the type of quadrature and we also want to capture the ordering of
+# the datatypes somehow (see operator:dim_info for example).
+array_datatypes = [
+    {"name": "operator", "scalar_type": "lfric real scalar",
+     "dims": ["ndf", "ndf", "ncell_3d"], "vars": ["fs_from", "fs_to"],
+     "dim_info": {"fs_from": 0, "fs_to": 1, "ncells": 2}},
+    {"name": "dof map", "scalar_type": "lfric integer scalar",
+     "dims": ["ndf"], "vars": ["fs"]},
+    {"name": "basis function", "scalar_type": "lfric real scalar",
+     "dims": [1, 1, 1], "vars": ["fs", "quad"]},
+    {"name": "diff basis function", "scalar_type": "lfric real scalar",
+     "dims": [1, 1, 1], "vars": ["fs", "quad"]},
+    {"name": "qr weights in horizontal", "scalar_type": "lfric real scalar",
+     "dims": ["number of qr points in horizontal"], "vars": []},
+    {"name": "qr weights in vertical", "scalar_type": "lfric real scalar",
+     "dims": ["number of qr points in vertical"], "vars": []},
+    {"name": "qr weights", "scalar_type": "lfric real scalar",
+     "dims": ["number of qr points"], "vars": []}]
+
+for array_type in array_datatypes + field_datatypes:
+    NAME = "".join(array_type["name"].title().split())
+    DIMS = array_type["dims"]
+    SCALAR_TYPE = "".join(array_type["scalar_type"].title().split())
+    ARGS = ["self", "name", "dims"] + array_type["vars"] + ["interface=None"]
+    VARS = ["        self.{0} = {0}".format(var) for var in array_type["vars"]]
+    exec('''
+class {0}DataType(ArrayType):
```

Comment:
Comments!

Reply:
As mentioned earlier, I've commented all of PSyIR but obviously let me know if you think more information should be added.